### PR TITLE
DuckType Best Practices in Readme.md

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR adds a best practices section to the DuckType readme file.

@DataDog/apm-dotnet

### CODE REVIEW

File: tracer/test/benchmarks/Benchmarks.Trace/DuckTyping/DuckTypeMethodCallComparisonBenchmark.cs [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+using Datadog.Trace.DuckTyping;
+using static Benchmarks.Trace.DuckTyping.ObscureObject;
+
+namespace Benchmarks.Trace.DuckTyping
+{
+    [DatadogExporter]
+    [MemoryDiagnoser]
+    [GroupBenchmarksBy(BenchmarkLogicalGroupRule.ByCategory)]
+    [CategoriesColumn]
+    public class DuckTypeMethodCallComparisonBenchmark
```
Comment (andrewlock) [inline tracer/test/benchmarks/Benchmarks.Trace/DuckTyping/DuckTypeMethodCallComparisonBenchmark.cs:17]: These are good to have - my only question is whether we want to run them all the time 🤔 There's a lot here, so they probably add a while to the benchmark run time?

**Review by andrewlock — APPROVED** [review]

File: tracer/test/benchmarks/Benchmarks.Trace/DuckTyping/DuckTypeMethodCallComparisonBenchmark.cs (hunk 1 above)
Comment (tonyredondo) [reply tracer/test/benchmarks/Benchmarks.Trace/DuckTyping/DuckTypeMethodCallComparisonBenchmark.cs:17]: Currently all DuckType benchmarks are being ignored.
https://github.com/DataDog/dd-trace-dotnet/blob/53c005ddb1925f765fda61879d73c00c95b87fab/tracer/test/benchmarks/Benchmarks.Trace/Benchmarks.Trace.csproj#L22-L29

Comment (andrewlock) [reply tracer/test/benchmarks/Benchmarks.Trace/DuckTyping/DuckTypeMethodCallComparisonBenchmark.cs:17]: Oh yeah 🤦‍♂️ Nice 👍

File: tracer/src/Datadog.Trace/DuckTyping/README.md [hunk 2]
```diff
@@ -340,11 +340,106 @@ public interface IProxyMyHandlerConfiguration
...
+│ Can the proxy be added │   Yes      │Use an Interface│
+│ as a constraint to the ├───────────►|     proxy      │
+│      integration?      │            └────────────────┘
+└───────────┬────────────┘
+            │ No
+            │
+┌───────────▼────────────┐   Yes      ┌────────────────┐
+│  Do you need to call   ├───────────►|Use an Interface│
+│ a method on the proxy? │            |      proxy     |    
+└───────────┬────────────┘            └────────────────┘
```
Comment (zacharycmontoya) [inline tracer/src/Datadog.Trace/DuckTyping/README.md:363]: I thought at this point we would want to use a class with virtual method definitions?

Comment (tonyredondo) [reply tracer/src/Datadog.Trace/DuckTyping/README.md:363]: Yes that was the previous advice, virtual method are a little bit faster than an interface method call. But I ran a new benchmark using the actual ducktype implementation and with all the options we have, and the difference is something like 1 or 2 nanoseconds, so I really think it doesn't worth.

This PR includes the results of the benchmarks and as a conclusion we should really try to use the generic constraints (20ns diff).

**Review by lucaspimentel — APPROVED** [review]

File: tracer/src/Datadog.Trace/DuckTyping/README.md (hunk 2 above)
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/DuckTyping/README.md:363]: Sounds good 👍🏼

**Review by zacharycmontoya — APPROVED** [review]

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #2223 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/3cafe9e99979ce268515dcf5354e4454c734c2ad) will **not change** line coverage
:heavy_check_mark: Merging #2223 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/3cafe9e99979ce268515dcf5354e4454c734c2ad) will **not change** branch coverage
:heavy_check_mark: Merging #2223 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/3cafe9e99979ce268515dcf5354e4454c734c2ad) will will **decrease** complexity by `25`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/3cafe9e99979ce268515dcf5354e4454c734c2ad) | #2223       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9218` / `12597` | `9183` / `12556` |          |
| Lines %   | `73%`      | `73%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4206` / `6160` | `4186` / `6132` |          |
| Branches %| `68%`      | `68%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7357`      | `7332`        |  `-25`  :heavy_check_mark:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84605&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84605/artifacts?artifactName=Code%20Coverage%20Report_84605&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2223](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84646&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84646/artifacts?artifactName=Code%20Coverage%20Report_84646&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/3cafe9e99979ce268515dcf5354e4454c734c2ad) | #2223       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `73%`      | `73%`       |  `0%` :heavy_check_mark:  |
| Branches %| `68%`      | `68%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7357`      | `7332`       |  `-25` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Couchbase.IIOServiceExecuteIntegrationBis](https://github.com/DataDog/dd-trace-dotnet/tree/f4bbf62c80e0d65464990425628b210468ac8063/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Couchbase/IIOServiceExecuteIntegrationBis.cs) | `-100%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.MsTestV2.TestMethodRunnerExecuteIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/f4bbf62c80e0d65464990425628b210468ac8063/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/TestMethodRunnerExecuteIntegration.cs) | `-33%` :no_entry: | `-19%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Ci.Agent.CIAgentWriter](https://github.com/DataDog/dd-trace-dotnet/tree/f4bbf62c80e0d65464990425628b210468ac8063/tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs) | `-10%` :no_entry: | `-21%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.NUnit.NUnitCompositeWorkItemSkipChildrenIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/f4bbf62c80e0d65464990425628b210468ac8063/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/NUnit/NUnitCompositeWorkItemSkipChildrenIntegration.cs) | `-8%` :no_entry: | `-8%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.SpanContextPropagator](https://github.com/DataDog/dd-trace-dotnet/tree/f4bbf62c80e0d65464990425628b210468ac8063/tracer/src/Datadog.Trace/SpanContextPropagator.cs) | `10%` :heavy_check_mark: | `12%` :heavy_check_mark:   | `-25` :heavy_check_mark: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84605&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84605/artifacts?artifactName=Code%20Coverage%20Report_84605&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2223](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84646&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84646/artifacts?artifactName=Code%20Coverage%20Report_84646&api-version=6.0&%24format=zip)

**Pull request merged**